src/Analyzer/CodeAnalyzer.cs

```diff
@@ -0,0 +1,171 @@
+        private readonly IConfigReader configReader;
+        private readonly IReporter reporter;
+        private readonly IConsoleTimer timer;
+
+        private IPluginHandler pluginHandler;
+        private ISqlFileProcessor fileProcessor;
+
+        private const string sqlExtension = ".sql";
+        private const string defaultTSQLLintConfiguration = ".tsqllintrc";
+
+        private readonly string tmpTSQLLintPath;
+
+        /// <summary>
+        /// Tool integration constructor.
+        /// This will prepare everything needed for the tool to work with codacy
+        /// integration, using the Seed.
+        /// </summary>
+        public CodeAnalyzer() : base(sqlExtension)
+        {
+            this.timer = new ConsoleTimer();
+            this.reporter = new CodacyConsoleReporter();
+            this.configReader = new ConfigReader(reporter);
+
+            // create temporary directory
+            var tmpTSQLLintFolder = Path.Combine(Path.GetTempPath(), "tsqllint_" + Guid.NewGuid());
+            Directory.CreateDirectory(tmpTSQLLintFolder);
+
+            this.tmpTSQLLintPath = Path.Combine(tmpTSQLLintFolder, defaultTSQLLintConfiguration);
+
+            var tsqllintConfig = new TSQLLintConfiguration();
+
+            if (!(PatternIds is null) && PatternIds.Any())
+            {
+                tsqllintConfig.Rules = new Dictionary<string, string>();
+                foreach (var pattern in CurrentTool.Patterns)
+                {
+                    tsqllintConfig.Rules.Add(pattern.PatternId, "error");
+                }
+
+                foreach (var unusedPattern in Patterns.Patterns.Select(p => p.PatternId)
+                    .Except(CurrentTool.Patterns.Select(p => p.PatternId)))
+                {
+                    tsqllintConfig.Rules.Add(unusedPattern, "off");
+                }
+            } else if (File.Exists(defaultTSQLLintConfiguration))
+            {
+                var tsqlliteJSON = File.ReadAllText(defaultTSQLLintConfiguration);
+                var currentTSQLLintConfig = JsonConvert.DeserializeObject<TSQLLintConfiguration>(tsqlliteJSON);
+                tsqllintConfig.Rules = currentTSQLLintConfig.Rules;
+                tsqllintConfig.CompatibilityLevel = currentTSQLLintConfig.CompatibilityLevel;
+            }
+            else
+            {
+                tsqllintConfig.Rules = new Dictionary<string, string>();
+                foreach (var pattern in Patterns.Patterns.Select(p => p.PatternId))
+                {
+                    tsqllintConfig.Rules.Add(pattern, "error");
+                }
+            }
+
+            if (tsqllintConfig.CompatibilityLevel is null)
+            {
+                tsqllintConfig.CompatibilityLevel = 120;
+            }
+
+            File.WriteAllText(tmpTSQLLintPath, tsqllintConfig.ToString());
+        }
+
+        /// <summary>
+        /// Free temporary resources created on object construction.
+        /// </summary>
+        ~CodeAnalyzer()

```

Comment:
About this, I already had a discussion with @lolgab . I can't use `using` in this case, because I'm writing the temporary config file to be used on the tool after (when running), and only after the tool finish analyzing, I delete the resources using `~CodeAnalyzer()` dtor.


Check the comment, here: https://github.com/codacy/codacy-sonar-csharp/pull/21
And also, the dtors/finalizers: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/destructors

Reply:
exactly, you _should not_ do this in the destructor, you do not have any guarantee that the GC will pass before the program exit.
Please find a way to use a `using` block (or an equivalent) (e.g. with `async` actions or Tasks ...) because this, unfortunately, doesn't work.

Please open a task for fixing this problem also in:
https://github.com/codacy/codacy-sonar-csharp/pull/21

Reply:
Ok. I searched about that, and I managed to make `CodeAnalyzer` as a Disposable object. So now I'm using `using` on the main and this will unsure that the resources are deleted.

I'll also create a task for sonar-csharp. Thanks for making me aware of this!

Reference to: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose

Reply:
Done